rpc: track failed connections

koorosh opened the pull request:
This is an attempt to track information about broken connections to 
have visibility of both healthy and broken connections. 
Timestamp at which connection was broken is tracked and periodically
it attempts to restore connection.

In addition to existing changes, following should be resolved:
- should we indefinitely retry to restore connection or give up 
after several attempts?
- If connection is failed because of for instance wrong address or Node ID, 
should we somehow detect such cases or not?

Jira epic: https://cockroachlabs.atlassian.net/browse/CRDB-21710

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/96566)
<!-- Reviewable:end -->

tbg commented on pkg/rpc/context.go:
```diff
@@ -665,6 +665,24 @@ func NewContext(ctx context.Context, opts ContextOptions) *Context {
 	// a connection class. Only calls going over an actual gRPC connection will
 	// use that interceptor.
 
+	rpcCtx.Stopper.RunAsyncTask(ctx, "reconnect peers", func(ctx context.Context) {
```
Also, and maybe this is making perfect the enemy of good, sort of sad to have a single goroutine that's always there and ticking but can then get stuck on an individual node.

We could try to introduce, on `peer`, an async probing-based circuit breaker (util/circuit) and make its probe loop responsible for reconnecting. This would be beneficial in its own right, since it would (later, and not related to your work directly) allow us to phase out the circuit breakers we have at the `nodedialer` level.

tbg reviewed (commented):
Flushing some comments.

Before we get wrapped up in the code changes too much, let's make sure this suggested behavior is reasonable:

- keep state for all nodes
- dial unreachable nodes periodically
- but not if they're decommissioned (source of truth is tombstone storage)
- remove tracked state after nobody has tried to dial a given node for 2h

koorosh commented on pkg/rpc/context.go:
```diff
@@ -2042,32 +2060,91 @@ func (rpcCtx *Context) GRPCDialPod(
+	defer m.mu.Unlock()
+	for c := range m.mu.m {
+		if c.nodeID == nodeID {
+			delete(m.mu.m, c)
```
Done. `decommissioned` flag is used instead of removing records from map.
Now, when connection fails, we derive information whether node is decommission based on error type that in turn based on `nodeTombstoneStorage` logic.

tbg commented on pkg/rpc/context.go:
```diff
@@ -2204,18 +2264,90 @@ func (rpcCtx *Context) grpcDialNodeInternal(
+						// Wait for heartbeat timeout to ensure that there's enough time
+						// for initial heartbeat and then check connection health.
+						<-time.After(rpcCtx.heartbeatTimeout)
+						err := c.Health()
```
Why sleep and then call `c.Health()` when you can call `.Connect`?

tbg commented on pkg/rpc/context.go:
```diff
@@ -2204,18 +2264,90 @@ func (rpcCtx *Context) grpcDialNodeInternal(
+						if !inserted {
+							// someone else inserted connection and probably
+							// started dialing.
+							continue
```
Why does it matter who inserted? We are just trying to make sure there's a Connection, and we want to see it in a healthy state (call `Connect`). Why the `continue` here?

tbg commented on pkg/rpc/context.go:
```diff
@@ -2204,18 +2264,90 @@ func (rpcCtx *Context) grpcDialNodeInternal(
+					defer done()
+					for {
+						// periodically run probe until connection is healed.
+						// TODO (koorosh): would it be enough to pause for `heartbeatInterval` duration?
```
https://cockroachlabs.slack.com/archives/C04G3B34PEZ/p1677253494714269

tbg commented on pkg/rpc/context.go:
```diff
@@ -2204,18 +2264,90 @@ func (rpcCtx *Context) grpcDialNodeInternal(
+	// TODO (koorosh): breaker is stored per `peer` and shouldn't be reinitialized every time `grpcDialNodeInternal`
+	// is called.
+	if pc.breaker == nil {
+		// TODO (koorosh): Probably this logic should be extracted from here...
```
+1 maybe to `rpc/breaker.go` because you'll ultimately have to disable the existing breaker functionality I think! That functionality mostly powers `(*nodedialer.Dialer).getBreaker` but now components that sit on top of `rpc.Context` don't need their own breaker any more, they'll just blindly dial and the `rpc.Context` will fail-fast as appropriate.

tbg commented on pkg/util/grpcutil/grpc_util.go:
```diff
@@ -105,6 +105,17 @@ func IsClosedConnection(err error) bool {
+// due to remote node has been decommissioned.
+func IsDecommissionedError(err error) bool {
+	if s, ok := status.FromError(errors.UnwrapAll(err)); ok {
+		if s.Code() == codes.FailedPrecondition && strings.Contains(err.Error(), "was permanently removed from the cluster") {
```
String matching is not OK here. We need to convince ourselfes that at the caller, `FailedPrecondition` uniquely communicates a decommissioned peer in the place in which we're getting the error. You're interested in the `OnOutgoingPing` interceptor, right? Luckily this is called directly (i.e. not underneath gRPC) so the structure of the error remains but even better, we can adjust the signature, so I think you can change the signature of `OnOutgoingPing` to something like `(decommissioned bool, err error)`:

```go
		OnOutgoingPing: func(ctx context.Context, req *rpc.PingRequest) (decommissioned bool, _ error) {
			// Outgoing ping will block requests with codes.FailedPrecondition to
			// notify caller that this replica is decommissioned but others could
			// still be tried as caller node is valid, but not the destination.
			err := checkPingFor(ctx, req.TargetNodeID, codes.FailedPrecondition)
			return err != nil, err
		},
```

Then you shouldn't need to add `IsDecommissionedError` here.

It would be nice to have a "globally unique" way to determine whether an error is about a decommissioned node, but last time I checked there was a lot of protobuf marshalling related madness due to `*grpc.Status`; let's not stir that pot now.

tbg replied:
Ah, in case it's not clear, there wouldn't be a method `IsDecommissionedError` in this package at the end of this.

tbg committed:
stop: add OnQuiesce

I've noticed in the past that this is needed sometimes, including
in the next commit.

Epic: none
Release note: None

tbg committed:
breaker: allow testing override of Signal

We'll need this to port existing tests that rely on manually manipulating
breaker state.

Epic: none
Release note: None

tbg committed:
rpc: small comment tweak

Epic: None
Release note: None

tbg committed:
rpc: fix incorrect comment

What it said used to be true, but it caused all kinds of problems and
so we've stopped sharing long ago.

Epic: None
Release note: None

tbg committed:
rpc: touch up DialBack functionality

Clarified the semantics a bit and made sure it didn't dig into the lower layers
more than it needed to. Left some comments around oddities of the approach that
are good to know when working in the area.

Epic: none
Release note: None

tbg committed:
kvpb: introduce NodeDecommissionedError

This will be used in a follow-up commit to more precisely communicate that a
connection attempt failed due to a decommissioned peer or source.

It will be more important to get this right as we switch to a stateful
connection pool. We don't want to attempt to reconnect to decommissioned nodes
forever, and we also don't want to misrepresent them as "unavailable" for the
purpose of network observability.

So far we're relying on the `PermissionDenied` error code but this is already
overloaded. It's better to provide an explicit gRPC status payload that leaves
no room for interpretation.

Epic: none
Release note: None

tbg committed:
rpc: fix display for remoteNode 0

It was printing byte '?' instead of "?" previously.

Epic: None
Release note: None

tbg committed:
circuit: provide type for the anonymous Signal interface

It will be convenient to refer to it instead of typing out the anonymous
interface multiple times. I'm not sure why I didn't introduce this much
earlier, must have been a long bout of misguided purism.

Epic: none
Release note: None

tbg committed:
rpc: rename a package

A type peer will be introduced soon.

Epic: None
Release note: None

tbg committed:
server: fix a (harmless) test buglet

Can't scan from `/Min`.

Epic: none
Release note: None

tbg committed:
server: add comment pointing out TestGraphite slowness

Epic: none
Release note: None

tbg committed:
rpc: minor Goland lint appeasement

Epic: none
Release note: None

tbg committed:
rpc: avoid shadowing `backoff` package

Epic: none
Release note: None

tbg committed:
rpc: make some unused args anonymous

Epic: none
Release note: None

tbg committed:
wip

tbg committed:
nodedialer: remove test for deprecated breaker from TestConnHealth{,TryDial}

Epic: none
Release note: None

tbg committed:
nodedialer: modernize TestDialNoBreaker

Test the functionality via the new circuit breakers, i.e. this test will
continue working if we remove the old breakers (which it no longer tests).

Epic: none
Release note: None

tbg committed:
nodedialer: remove TestConcurrentCancellationAndTimeout

It was specific to the nodedialer-level circuit breakers, which will be
removed.

Epic: none
Release note: None

tbg committed:
nodedialer: remove breaker check from TestNodedialerPositive

This was testing the deprecated breaker.

Epic: none
Release note: None

tbg committed:
nodedialer: remove Test{Disconnects,ResolverErrors}Trip

Epic: none
Release note: None

tbg committed:
importer: use new rpc breaker in TestImportRetriesBreakerOpenFailure

Epic: none
Release note: None

tbg committed:
nodedialer: remove use of old breaker in TestConnHealthInternal

Epic: none
Release note: None

tbg committed:
rafttransport use new breakers and rework stuff, still need to take out one call to old breaker

tbg committed:
base: set explicit small RPCHeartbeatInterval

The previous default, zero, effectively implied a tight loop of heartbeats. This can't be good for race/stress builds, so add at least a little breath.

Epic: None
Release note: None

tbg commented:
Closing for https://github.com/cockroachdb/cockroach/pull/99191, which I'm actively working on.
